### PR TITLE
Implement NEMO Interface

### PR DESCRIPTION
**Pull request by matthewlh**

### CODE REVIEW

File: communications/command_definitions.py [hunk 1]
```diff
@@ -441,3 +441,27 @@ def insert_line_in_file(self, **kwargs):
...
+        self.parent.nemo_manager.set_config(kwargs)
+
+    def nemo_power_off(self):
+        self.parent.nemo_manager.power_off()
+
+    def nemo_power_on(self):
+        self.parent.nemo_manager.power_on()
+
+    def nemo_reboot(self):
+        self.parent.nemo_manager.reboot()
```
Comment (matthewlh) [inline communications/command_definitions.py:467]: @tmf97 Which flight mode(s) should these commands be added to? Just `normal_comands` or `command_commands`?

Comment (tmf97) [reply communications/command_definitions.py:467]: I'd say normal mode. Command mode is more meant for either timing-critical command execution or a sort of manual control over the spacecraft

**Review by tmf97 — CHANGES_REQUESTED** [review]
Looks mostly good from what I see. 
I would be happy to test this on our HITL table, although I'm not sure how to fully test all aspects of the nemo driver and manager, and the relevant commands. Is there a procedure you would like me to follow?
One thing that is missing is the command argument packing/unpacking. See the NormalMode class in `flight_mode.py`  for an example on how to use that flight mode's `command_codecs` and `command_arg_types` dictionaries to define how your arguments are packed/unpacked.

File: drivers/nemo/nemo.py [hunk 2]
```diff
@@ -0,0 +1,701 @@
...
+        15: '',
+    }
+
+    ASSEMBLY_DET_DICT = {
+        'ETU1': {'det0_mfg_serial_number': '01854', 'det1_mfg_serial_number': '01678'},
+        'FU1': {'det0_mfg_serial_number': '01867', 'det1_mfg_serial_number': '01898'},
+        'FU2': {'det0_mfg_serial_number': '00581', 'det1_mfg_serial_number': '00745'},
+    }
+
+    def __init__(self, dev_addr=0x13, reset_gpio_ch=5, log=True):
```
Comment (tmf97) [inline drivers/nemo/nemo.py:333]: Please change reset_gpio_ch to 16. GPIO pin 5 is not connected to anything on the flight units.

**Review by tmf97 — CHANGES_REQUESTED** [review]

**Comment by tmf97** [comment]
From the Oxygen flight unit, and running main on the nemo branch gives the following logs:
```pi@cislunar-oxygen:~/FlightSoftware $ python3 main.py 
2021-03-24 17:21:41,534 [INFO] [main.py:48] Initializing...
2021-03-24 17:21:41,671 [INFO] [main.py:124] Gom initialized
2021-03-24 17:21:41,894 [INFO] [main.py:132] Gyro initialized
2021-03-24 17:21:41,925 [INFO] [main.py:141] ADC initialized
2021-03-24 17:21:41,939 [INFO] [main.py:149] RTC initialized
2021-03-24 17:21:41,996 [ERROR] [nemo_manager.py:220] Nemo I2C transaction failure in NemoManager.set_config
2021-03-24 17:21:42,009 [INFO] [main.py:157] NEMO initialized
2021-03-24 17:21:42,016 [ERROR] [nemo_manager.py:121] Nemo I2C transaction failure in NemoManager tread
2021-03-24 17:21:42,060 [INFO] [ax5043_manager.py:105] Initializing
2021-03-24 17:21:42,113 [INFO] [ax5043_manager.py:109] POST passed
2021-03-24 17:21:42,121 [INFO] [main.py:165] Radio initialized
2021-03-24 17:21:42,141 [INFO] [main.py:175] Mux initialized
2021-03-24 17:21:42,261 [INFO] [restart_reboot.py:60] Restarting...
2021-03-24 17:21:43,049 [ERROR] [nemo_manager.py:121] Nemo I2C transaction failure in NemoManager tread
2021-03-24 17:21:44,070 [ERROR] [nemo_manager.py:121] Nemo I2C transaction failure in NemoManager tread
2021-03-24 17:21:45,083 [ERROR] [nemo_manager.py:121] Nemo I2C transaction failure in NemoManager tread
2021-03-24 17:21:46,094 [ERROR] [nemo_manager.py:121] Nemo I2C transaction failure in NemoManager tread
2021-03-24 17:21:47,104 [ERROR] [nemo_manager.py:121] Nemo I2C transaction failure in NemoManager tread
2021-03-24 17:21:48,116 [ERROR] [nemo_manager.py:121] Nemo I2C transaction failure in NemoManager tread
2021-03-24 17:21:49,127 [ERROR] [nemo_manager.py:121] Nemo I2C transaction failure in NemoManager tread
2021-03-24 17:21:49,994 [INFO] [ax5043_manager.py:125] Autoranging
2021-03-24 17:21:50,027 [INFO] [flight_mode.py:197] Starting flight mode 1
2021-03-24 17:21:50,151 [ERROR] [nemo_manager.py:121] Nemo I2C transaction failure in NemoManager tread
2021-03-24 17:21:51,175 [ERROR] [nemo_manager.py:121] Nemo I2C transaction failure in NemoManager tread
2021-03-24 17:21:51,677 [INFO] [flight_mode.py:201] Finishing flight mode 1
```

Running `i2cdetect -y -a 1` shows address 0x13 is detected, although we're getting I2C transaction failures. It looks like everything is wired up correctly, but maybe we're running into some thread-related I2C issues? (every component except for the radio uses i2c)

File: drivers/nemo/nemo.py (hunk 2 above)
Comment (matthewlh) [reply drivers/nemo/nemo.py:333]: Changed in 65ecd968fc32a42b09db3b625101de1d0d523a6a

**Comment by matthewlh** [comment]
The I2C transaction failures may be due to the reset line issue, I've resolved that in 65ecd968fc32a42b09db3b625101de1d0d523a6a. Can you give it a try with that change?

**Comment by tmf97** [comment]
I replaced 5 with 16 in the nemo init before testing on the flight unit, which I believe should have the same effect as the change you just pushed.

**Comment by matthewlh** [comment]
Hmm, where you running `i2cdetect -y -a 1` while main.py was running? If not, could you try?

That error being logged means `self._bus.writeto*` is failing in either `Nemo._read_register()` or `Nemo._write_register()` with `OSError(errno=121)`.

**Comment by tmf97** [comment]
I ran `i2cdetect` before running main. Does `errno=121` mean an electrical connection is not being made?
I can test running the `i2cdetect` command while running main on the flight units later today. It might be worthwhile for me to try and initialize a `NemoManager` on our HITL table (where I trust the wiring more than on the flight unit)
Also, for purposes of making sure that everything is wired correctly, I've been assuming that if I see `NEMO initialized` that all wired connections are good. Is this a OK assumption to make?

**Comment by matthewlh** [comment]
`errno=121` means the device (0x13) NACK'd (or no device ACK'd) the transaction. So, I suspect `i2cdetect` will not see 0x13 while main.py is running.

Sure, testing this branch on the HITL would be good.

`NEMO initialized` simply means the NemoManager constructor ran to completion. It does not indicate correct hardware connections. `set_config` is called from there and is failing. It's setup this way so if a transaction fails during the constructor, the thread is still started and can recover if NEMO starts responding on the bus in the future.

`sudo picberry --gpio=20,21,16 --family=pic24fjxxxgb2xx --blankcheck` is a good way to determine if all the non-I2C lines are connected correctly. Have you installed picberry on the flight units?

`i2cdetect` showing 0x13 means that at least at that moment, Power, Ground, SDA, and SCL were likely connected correctly to NEMO (assuming nothing else will respond to 0x13).

**Comment by tmf97** [comment]
Thanks for the info!
Everything worked well on the HITL, so there's something up with the flight unit wiring.
We do not have picberry installed on the flight units, but I can install it to check wiring.

**Comment by tmf97** [comment]
After a little bit of working on Oxygen, we can successfully talk with the NEMO flight unit.
I did not install picberry on the pi. Would you need it to be on the pi's during flight?

**Review by tmf97 — APPROVED** [review]

**Comment by tmf97** [comment]
Since initializing the NemoManager works on both the HITL and Oxygen, I'm approving these changes. We can integrate the NEMO commands into the main flight software in a new PR.

**Pull request merged**